libcnb-test/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+    /// References this buildpack in the Rust Crate under test
+    Crate,
+    /// References another buildpack by id, local directory or tarball
+    Other(String),
+}
+
+impl IntegrationTest {
+    /// Creates a new integration test.
+    ///
+    /// # Panics
+    /// - When the connection to Docker failed
+    /// - When the internal Tokio runtime could not be created
+    pub fn new(builder_name: impl Into<String>, app_dir: impl AsRef<Path>) -> Self {
+        IntegrationTest {
+            app_dir: PathBuf::from(app_dir.as_ref()),
+            target_triple: String::from("x86_64-unknown-linux-musl"),
+            builder_name: builder_name.into(),
+            buildpacks: vec![BuildpackReference::Crate],
+            docker: Docker::connect_with_local_defaults()
+                .expect("Could not connect to local Docker deamon"),
+            tokio_runtime: tokio::runtime::Runtime::new()
+                .expect("Could not create internal Tokio runtime"),
+        }
+    }
+
+    /// Sets the buildpacks order.
+    ///
+    /// Defaults to [`BuildpackReference::Crate`].
+    pub fn buildpacks(&mut self, buildpacks: impl Into<Vec<BuildpackReference>>) -> &mut Self {
+        self.buildpacks = buildpacks.into();
+        self
+    }
+
+    /// Sets the target triple.
+    ///
+    /// Defaults to `x86_64-unknown-linux-musl`.
+    pub fn target_triple(&mut self, target_triple: impl Into<String>) -> &mut Self {
+        self.target_triple = target_triple.into();
+        self
+    }
+
+    /// Starts a new integration test run.
+    ///
+    /// This function will copy the application to a temporary directory, cross-compiles this crate,
+    /// packages it as a buildpack and then invokes [pack](https://buildpacks.io/docs/tools/pack/)
+    /// to build a new Docker image with the buildpacks specified by this integration test instance.
+    ///
+    /// Since this function is supposed to only be used in integration tests, failures are not
+    /// signalled via [`Result`](Result) values. Instead, this function panics whenever an unexpected error
+    /// occurred to simplify testing code.
+    ///
+    /// # Panics
+    /// - When the app could not be copied
+    /// - When this crate could not be packed as a buildpack
+    /// - When the "pack" command unexpectedly fails
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libcnb_test::IntegrationTest;
+    ///
+    /// IntegrationTest::new("heroku/buildpacks:20", "test-fixtures/app")
+    ///     .run(|context| {
+    ///         assert!(context.pack_stdout.contains("---> Ruby Buildpack"));
+    ///         assert!(context.pack_stdout.contains("---> Installing bundler"));
+    ///         assert!(context.pack_stdout.contains("---> Installing gems"));
+    ///     })
+    /// ```
+    pub fn run<F: FnOnce(IntegrationTestContext)>(&mut self, f: F) {
+        let temp_app_dir =
+            app::copy_app(&self.app_dir).expect("Could not copy app to temporary location");
```

Comment:
For the classic Python buildpack, I've favoured having explicit test fixtures (https://github.com/heroku/heroku-buildpack-python/tree/main/spec/fixtures) over programmatically generating them, since:
- it makes it easier to run equivalent builds outside of the tests (in the case of CNBs, running `pack build` manually)
- it allows tooling to operate on the app source (eg syntax highlighting, Dependabot, linting etc)
- it's faster than having to make temporary copies all the time

Now there are clearly cases where it is useful (or necessary) to mutate the app source locally - eg when testing cache-busting or similar via repeat builds. However I wonder if we should at least support skipping the copy for use-cases where nothing is being written during the test? (I'm fine with this being a GitHub issue for later, if we think supporting this is worthwhile)

Reply:
I agree about using explicit test fixtures. This is a defensive measure, especially with the concurrent execution that would, in case of modifications, lead to hard to debug bugs.

In some cases I can see us modifying explicit test fixtures during the test, for example asserting that runtime version switching works for the same app with a warm cache.

I wouldn't even add an issue for this a this point and cross the bridge when we get there. I wouldn't object if you'd add an issue for this though. :)